### PR TITLE
feat(diagram): add generated ERD diagram

### LINKED ISSUE #340: Database diagram support
1. Extend the handling of fenced block languages to include "database"
2. Parse the content of the fenced block as YAML
3. Use the parsed values to extract schema information from a live database and generate source for a Mermaid entity relationship diagram. 
4. Render the generated Mermaid diagram as SVG
5. Based on the method used for Kroki diagrams add caching

For step 2, parse YAML like so. `token.text` gives you everything between the fences.

``` typescript

          case "PUSH":
            resolvedConfig = yaml.parse(token.text);
            stack.push(resolvedConfig);
            break;
```

The YAML parser is already in scope unless you break out the logic into a new file in which case you can import like so:

``` typescript 
import * as yaml from "yaml";
```

## Sample

````
```database
DatabaseType: postgresql
ConnectionString: Host=localhost;Port=5432;Database=mydatabase;Username=user;Password=secret
Schema: public
Tables: 
  - Offers
  - Orders
  - Passes
Detail: column names
```
````

## DatabaseType

- postgresql
- mssql

## ConnectionString

A dotnet database connection string. Depending on your setup you may need to embed the password. In that case create a user with permissions limited to reading the schema details.

## Schema

The schema containing the tables to be diagrammed. When this item is omitted, the Postgres default is `public` and the MSSQL default is `dbo`.

## Tables

A list of the tables you want in the diagram. When this item is omitted, the default is all the tables.

## Detail
Supported values in decreasing level of detail are

- column types
- column names
- key column names
- table name

When this item is omitted, the default is column names.

### PR DESCRIPTION
**Pull request by abdiukov**
This adds a new fenced block language called `database`.

<pre>
```database 
DatabaseType: mssql
ConnectionString: Server=localhost,1433; Database=OrderManagementDb; User Id=sa; Password=yourStrong(!)Password;
Schema: dbo
Tables: 
  - OrderStates
  - OrderHeaders
  - OrderItems
  - StockItems
Detail: keys
```
</pre>

From that, it draws a diagram, as can be seen below:

image

More work is remaining! We need to add unit tests and error handling, allowing users to see why the diagram failed to draw. I also noticed that if not all tables are provided, the diagram rendering gets borked.

Partially resolves #340, more work is remaining.

### CODE REVIEW

File: dist/.gitkeep
Comment (abdiukov) [inline dist/.gitkeep:1]: I noticed on my machine that if `dist/` file is not present, it fails to build. 

As in, it wants a `dist` folder to be there **before ** the build.

File: doc/dev-notes.md [hunk 1]
```diff
@@ -1,3 +1,7 @@
+# What version of node is recommended?
+
+To run this in development environment, node version 18 is recommended.
```
Comment (abdiukov) [inline doc/dev-notes.md:3]: Could be useful for devs. But what do you think? What would be the best place to put this information?

**Comment by PeterWone** [comment]
Specify version 20 and leave it in dev-notes

File: doc/dev-notes.md (hunk 1 above)
Comment (PeterWone) [reply doc/dev-notes.md:3]: Specify version 20 and leave it in dev-notes

File: dist/.gitkeep
Comment (PeterWone) [reply dist/.gitkeep:1]: We'll have to modify the build process to probe and create it

**Comment by PeterWone** [comment]
> We need to add unit tests and error handling, allowing users to see why the diagram failed to draw.

For a machine-generated diagram like this, failures are restricted to things like network or credential failures. Instead of generating SVG, continue to generate an HTML token but return a div containing an error message. 

```
Unable to communicate with database. Manually check that you can connect to the database with the supplied credentials.
```

Give the div a CSS class. We can use this to style it later.

**Comment by abdiukov** [comment]
Checked that it functions for Postgres and MySQL, and updated the db.md document. Cheers.

image
image

File: doc/dev-notes.md (hunk 1 above)
Comment (abdiukov) [reply doc/dev-notes.md]: Good idea, specified the version 20

**Comment by abdiukov** [comment]
Error handling has been added. If the connection is unsuccessful, the following is displayed:
image

**Comment by abdiukov** [comment]
Great idea, I've changed the error handling. Now it looks like this. What do you think?
image

**Comment by abdiukov** [comment]
>  Is this because you only declared one FK while setting up your sample db?

Yes, good catch. I have only set up one foreign key relationship in that example there.

Here is the same PG table with more relationships. Cheers.
image

**Pull request merged**